integrations/wagmi-v2/src/index.ts

```diff
@@ -135,79 +135,108 @@ export function createSapphireConfig<
-										detail: {
-											...announceProviderEvent.detail,
-											provider: wrapEthereumProvider(
-												announceProviderEvent.detail.provider,
-												wrapOptions,
-											),
+							const {
+								info: { name, rdns, icon, uuid },
+								provider,
+							} = announceProviderEvent.detail;
+							const dispatchAnnounceProviderEvent: CustomEvent<EIP6963ProviderDetail> =
+								new CustomEvent(EIP6963_ANNOUNCE_PROVIDER_EVENT_NAME, {
+									detail: Object.freeze({
+										info: {
+											uuid: `sapphire.${uuid}`,
+											rdns: `sapphire.${rdns}` as Rdns,
+											name: `${name} (Sapphire)`,
+											icon,
 										},
-									},
-								);
-								/* Duplicate supported providers(replaceProviders = false), i.e. MetaMask -> [MetaMask, SapphireMetaMask] */
-							} else if (
-								wrappedProvidersFilter(announceProviderEvent.detail.info.rdns)
-							) {
-								const {
-									info: { name, rdns, icon, uuid },
-									provider,
-								} = announceProviderEvent.detail;
-
-								const dispatchAnnounceProviderEvent: CustomEvent<EIP6963ProviderDetail> =
-									new CustomEvent(EIP6963_ANNOUNCE_PROVIDER_EVENT_NAME, {
-										detail: Object.freeze({
-											info: {
-												uuid: `sapphire.${uuid}`,
-												rdns: `sapphire.${rdns}` as Rdns,
-												name: `${name} (Sapphire)`,
-												icon,
-											},
-											provider: wrapEthereumProvider(provider, wrapOptions),
-										}),
-									});
-
-								window.dispatchEvent(dispatchAnnounceProviderEvent);
-							}
+										provider: wrapEthereumProvider(provider, wrapOptions),
+									}),
+								});
+							window.dispatchEvent(dispatchAnnounceProviderEvent);
 						}
+					}
+
+					return typeof callback === "function"
+						? callback(patchCustomEvent ?? event)
+						: callback?.handleEvent(patchCustomEvent ?? event);
+				};
+
+				if (callback) {
+					// Store mapping between original callback and wrapper
+					listenerMap.set(callback, wrapperFn);
+				}
 
-						return typeof callback === "function"
-							? callback(patchCustomEvent ?? event)
-							: callback?.handleEvent(patchCustomEvent ?? event);
-					},
-					options,
-				);
+				_addEventListener.call(this, type, wrapperFn, options);
 			} else {
 				_addEventListener.call(this, type, callback, options);
 			}
 		},
 	});
 
+	Object.defineProperty(EventTarget.prototype, "removeEventListener", {
+		value: function (
+			type: string,
+			callback: EventListenerOrEventListenerObject | null,
+			options?: EventListenerOptions | boolean,
+		): void {
+			if (
+				type === EIP6963_ANNOUNCE_PROVIDER_EVENT_NAME &&
+				callback &&
+				listenerMap.has(callback)
+			) {
+				const wrapperFn = listenerMap.get(callback);
+
+				listenerMap.delete(callback);
+
+				_removeEventListener.call(this, type, wrapperFn ?? null, options);
+			} else {
+				_removeEventListener.call(this, type, callback, options);
+			}
+		},
+	});
```

Comment:
This seems annoying to maintain.

Can we reimplement the whole thing, and instead of patching
`addEventListener if eip6963:announceProvider wrapEthereumProvider`
could patch
`window.dispatchEvent if eip6963:announceProvider wrapEthereumProvider`